### DISCUSSION TITLE
watcher with Event?

### DISCUSSION THREAD

**Discussion by f4tq** [kube-rs/kube#1208] [General]
Hi,
I need to process my Resource events  Event::Added/Deleted which I can do with plain `my_api.watch`.  However, I'd like to use the reconnect handling the `watcher` interface provides but it strips the event and just returns the underlying Api resource with is useless to me without the event context.

Currently, the watch interface works but disconnects on a timeout which I don't want (I want it to run forever)
```
let proxy_api = Api::<FooProxy>::all(client);
 let mut watcher = proxy_api.watch(&ListParams::default(), "0").await?.boxed();
 while let Some(event) = watcher.next().await {
       match event {
          ... process
   }
}
```
but I'd like `watcher.next().await` to not timeout.  I'd put a loop around it but I hope to not see events repeated.
Is there another way to use watcher. 
For completeness, while this 'work', it strips the Event:
```
let ew = watcher(proxy_api, ListParams::default()).touched_objects();
while let Some(event: Foo) = ew.try_next().await? {
   /// missing event
}
```

Please advise how I can make either way work reliably with the goal of retries/reconnects and the actual event.
Thanks

**Comment by Dav1dde** [comment]
This doesn't work for you?:

```rs
let ew = watcher(proxy_api, ListParams::default()); // No `.touched_objects()`
```

**Comment by f4tq** [comment]
Doh.  I guess I got thrown the need to pin so I reached for the event flatteners. 
This fixed me:
```
        let wat = watcher(proxy_api, ListParams::default());
        let mut pinned_wat = Box::pin(wat);
        while let Some(event) = pinned_wat.try_next().await? {
           match event {
                Event::Applied(obj) => { .. }
...
```
Thanks for the help.  Works like a charm now.

**Reply by Dav1dde** [reply]
Btw you can pin on the stack with the [`pin_mut`](https://docs.rs/futures/latest/futures/macro.pin_mut.html) macro
